File: 06_LinkedList/Single Linked List/SLL CPP/SLL.cpp
#include "SLL.hpp"

SLL::SLL()
{
	head = nullptr;
}
	
SLL::SLL(Node* h)
{
	head = h;
}

void SLL::printList()
{
	if(head == nullptr)
	{
		cout<<"No List\n";
		return;
	}
	Node* temp  = head;
	while(temp->next != nullptr)
	{
		cout<<temp->val<<"--->";
		temp = temp->next;
	}
	cout<<temp->val;
	cout<<endl;
}
	
void SLL::addEnd(int newVal)
{
	Node* nn = new Node(newVal);
	
	//if there is no list
	if(head == nullptr)
	{
		head = nn;
		return;
	}
		
	//SLL traversal - add at the end of the list
	Node* temp = head;
	while(temp->next != nullptr)
	{
		temp = temp->next;
	}

	temp->next = nn;
}

void SLL::addAfter(int newVal, int prevVal)
{
	Node* nn = new Node(newVal);
	
	if(head == nullptr)
		return;
	
	Node* temp = head;
	while(temp->next != nullptr && temp->val != prevVal)
	{
		temp = temp->next;
	}
	
	if(temp == nullptr)
	{
		cout<<"Not Present"<<endl;
		return;
	}
	
	Node* newNext = temp->next;
	temp->next = nn;
	nn->next = newNext;
}

int SLL::deleteFirstInstance(int target)
{
	Node* temp = head;
	Node* prev;
	int retVal;
	if(head == nullptr)
	{
		return -9999;
	}

	while(temp != nullptr)
	{
		if(temp->val == target)
		{
			retVal = temp->val;
			if(temp == head)
			{
				head = head->next;
				break;
			}
			prev->next = temp->next;
			break;
		}
		prev = temp;
		temp = temp->next;
	}
	return retVal;
}

void SLL::deleteAllInstances(int target)
{
	Node* temp = head;
	Node* prev;
	Node* toDelete;

	if(head == nullptr)
	{
		return;
	}

	while(temp != nullptr)
	{
		if(temp->val == target)
		{
			toDelete = temp;
			if(temp == head)
			{
				head = head->next;
			}
			else
			{
			prev->next = temp->next;
			}
		}
		prev = temp;
		temp = temp->next;
	}
}

void SLL::deleteFront()
{
	if(head == nullptr)
	{
		return;
	}

	head = head->next;
}

void SLL::deleteEnd()
{
	if(head == nullptr)
	{
		return;
	}

	Node* temp = head;
	Node* prev;
	while(temp->next != nullptr)
	{
		prev = temp;
		temp = temp->next;
	}
	prev->next = nullptr;
}

void SLL::swap(Node*a, Node*b)
{
	int temp = a->val;
	a->val = b->val;
	b->val = temp;
}

void SLL:: sortSLL()
{
	if(head == nullptr)
	{
		return;
	}

	Node* i = head;
	Node* j;
	while(i != nullptr)
	{
		j = i->next;
		while (j != nullptr)
		{
			if(i->val > j->val)
			{
				swap(i, j);
			}
			j = j->next;
		}
		i = i->next;
	}

}

int main()
{
	cout<<"The Single Liked List\n";
    	SLL obj;
	obj.addEnd(7);
	obj.addEnd(9);
	obj.addEnd(10);
	obj.addEnd(1);
	obj.addEnd(0);
	obj.addEnd(9);
	obj.addEnd(2);
	obj.addEnd(4);
	obj.printList();

	cout<<"Add 10 after 7\n";
	obj.addAfter(10,7);
	obj.printList();

	cout<<"Add 3 after 9\n";
	obj.addAfter(3,9);
	obj.printList();

	cout<<"Sort List\n";
	obj.sortSLL();
	obj.printList();

	cout<<"Delete front\n";
	obj.deleteFront();
	obj.printList();

	cout<<"Delete end\n";
	obj.deleteEnd();
	obj.printList();

	cout<<"Delete the first instance of 9\n";
	obj.deleteFirstInstance(9);
	obj.printList();

	cout<<"Delete all instances of 10\n";
	obj.deleteAllInstances(10);
	obj.printList();
    
    	return 0;
}

File: 06_LinkedList/Single Linked List/SLL Java/SLL.java
import java.io.*;
import java.util.*;

public class SLL {
    private NodeSLL head;

    public SLL() {
        head = null;
    }

    public SLL(NodeSLL h) {
        head = h;
    }

    public void printList() {
        if (head == null) {
            System.out.println("No List");
            return;
        }

        NodeSLL temp = head;
        while (temp.next != null) {
            System.out.print(temp.val + "--->");
            temp = temp.next;
        }
        System.out.println(temp.val);
    }

    public void swap(NodeSLL a, NodeSLL b) {
        int temp = a.val;
        a.val = b.val;
        b.val = temp;
    }

    public void addEnd(int newVal) {
        NodeSLL nn = new NodeSLL(newVal, null);

        // if there is no list
        if (head == null) {
            head = nn;
            return;
        }

        // SLL traversal - add at the end of the list
        NodeSLL temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }

        temp.next = nn;
    }

    public void addAfter(int newVal, int prevVal) {
        NodeSLL nn = new NodeSLL(newVal, null);

        if (head == null)
            return;

        NodeSLL temp = head;
        while (temp.next != null && temp.val != prevVal) {
            temp = temp.next;
        }

        NodeSLL newNext = temp.next;
        temp.next = nn;
        nn.next = newNext;
    }

    public void deleteFirstInstance(int target) {
        NodeSLL temp = head;
        NodeSLL prev = null;
        int retVal;
        if (head == null) {
            return;
        }

        while (temp != null) {
            if (temp.val == target) {
                retVal = temp.val;
                if (temp == head) {
                    head = head.next;
                    break;
                }
                prev.next = temp.next;
                break;
            }
            prev = temp;
            temp = temp.next;
        }
    }

    public void deleteAllInstances(int target) {
        NodeSLL temp = head;
        NodeSLL prev = null;

        if (head == null) {
            return;
        }

        while (temp != null) {
            if (temp.val == target) {
                if (temp == head) {
                    head = head.next;
                } else {
                    prev.next = temp.next;
                }
            }
            prev = temp;
            temp = temp.next;
        }
    }

    public void deleteFront() {
        if (head == null) {
            return;
        }

        head = head.next;
    }

    public void deleteEnd() {
        if (head == null) {
            return;
        }

        NodeSLL temp = head;
        NodeSLL prev = null;
        while (temp.next != null) {
            prev = temp;
            temp = temp.next;
        }
        prev.next = null;
    }

    public void sortSLL() {
        if (head == null) {
            return;
        }

        NodeSLL i = head;
        NodeSLL j;
        while (i != null) {
            j = i.next;
            while (j != null) {
                if (i.val > j.val) {
                    swap(i, j);
                }
                j = j.next;
            }
            i = i.next;
        }
    }
}

class MainSLL {
    public static void main(String args[]) throws IOException {
        System.out.println("The Single Linked List");
        SLL obj = new SLL();
        obj.addEnd(7);
        obj.addEnd(9);
        obj.addEnd(10);
        obj.addEnd(1);
        obj.addEnd(0);
        obj.addEnd(9);
        obj.addEnd(2);
        obj.addEnd(4);
        obj.printList();

        System.out.println("Add 10 after 7");
        obj.addAfter(10, 7);
        obj.printList();

        System.out.println("Add 3 after 9");
        obj.addAfter(3, 9);
        obj.printList();

        System.out.println("Sorted List");
        obj.sortSLL();
        obj.printList();

        System.out.println("Delete front");
        obj.deleteFront();
        obj.printList();

        System.out.println("Delete end");
        obj.deleteEnd();
        obj.printList();

        System.out.println("Delete the first instance of 9");
        obj.deleteFirstInstance(9);
        obj.printList();

        System.out.println("Delete all instances of 10");
        obj.deleteAllInstances(10);
        obj.printList();
    }
}
